src/models/UniqueNamedContainer.java

```diff
@@ -41,63 +43,54 @@ public UniqueNamedContainer() {
 
     /**
      * Adds the specified element to the end of this container.
-     *
+     * <p>
      * For non singleton items renaming of its unique name will happen when:
-     *   There is another item with the same owner and the same original name.
-     *   There is another item with a different owner and the same original name.
+     * There is another item with the same owner and the same original name.
+     * There is another item with a different owner and the same original name.
      *
-     * @param item Item to be added to the end of this container.
+     * @param item item to be added to the end of this container.
      */
     public void add(T item) {
         T newItem = (T) item.getCopy();
 
-        if (item.isGlobal()) {
-            // We only want to add global items it if its unique name is not present.
-            Optional<T> existing = findFirstWithUniqueName(newItem.getUniqueName());
-            if (existing.isEmpty()) {
-                items.add(newItem);
-            }
-        } else {
-            // Unique name naming rules:
-            //   Same owner and different name: keep it as is.
-            //   Different owner and name: keep it as is.
-            //   Different or the same owner and the same original name: add owner to name "owner.n.name" where n is a counter value.
-            // Motivation: Machine <= Machine || Machine. Here the owner is Machine but their clocks are different.
-            List<T> sameOriginalName = items.stream()
-                    .filter(current -> Objects.equals(current.getOriginalName(), newItem.getOriginalName()))
+        if (!item.isGlobal()) {
+            List<T> sameName = items.stream()
+                    .filter(it -> sameName(it, item))
                     .collect(Collectors.toList());
-
-            if (!sameOriginalName.isEmpty()) {
-
-                List<T> sameOwner = sameOriginalName.stream()
-                        .filter(current -> Objects.equals(current.getOwnerName(), newItem.getOwnerName()))
-                        .collect(Collectors.toList());
-
-                if (sameOwner.size() == 1) {
-                    sameOwner.get(0).setUniqueName(1);
-                }
-
-                if (!sameOwner.isEmpty()) {
+            if (sameName.size() != 0) {
+                List<T> sameOwner = sameName.stream().filter(c -> c.getOwnerName().equals(item.getOwnerName())).collect(Collectors.toList());
+                if (sameOwner.size() != 0) { // Same name, same owner
+                    for (int i = 0; i < sameOwner.size(); i++) {
+                        sameOwner.get(i).setUniqueName(i + 1);
+                    }
```

Comment:
In this case, you found one or more items with the same name and same owner as the item you want to add. When this happens for the first time for this `uniqueNamedContainer`, you indeed have to rename all the entities. But the second time this happens, all the currently present items already have a unique name. So why do you need to rename those again?

Reply:
The second time we have to set the unique name of the first element such that it has the index in its name. However, all the other times we don't have to do this. Nice catch, I have fixed it and added a comment.